phd-tests/framework/src/serial/raw_buffer.rs

```diff
@@ -82,62 +106,37 @@ impl Inner {
     }
 }
 
-impl Drop for Inner {
-    fn drop(&mut self) {
-        if let Err(e) = self.log.flush() {
-            error!(%e, "failed to flush serial console log during drop");
-        }
-    }
-}
-
-impl vte::Perform for Inner {
-    fn print(&mut self, c: char) {
-        self.push_character(c);
-    }
-
-    fn execute(&mut self, byte: u8) {
-        if byte == b'\n' {
-            self.push_character('\n');
-        }
-    }
-}
-
-/// A "raw" serial console buffer that handles incoming characters and newline
-/// control bytes and nothing else.
-pub(super) struct RawBuffer {
-    inner: Inner,
-    parser: vte::Parser,
-}
-
-impl RawBuffer {
-    /// Constructs a new buffer.
-    pub(super) fn new(log_path: Utf8PathBuf) -> Result<Self> {
-        let log_file = std::fs::File::create(&log_path).with_context(|| {
-            format!("opening serial console log file {}", log_path)
-        })?;
-        let writer = BufWriter::new(log_file);
-        let inner = Inner {
-            log: writer,
-            line_buffer: String::new(),
-            wait_buffer: String::new(),
-            waiter: None,
-        };
-        Ok(Self { inner, parser: vte::Parser::new() })
-    }
-}
-
 impl Buffer for RawBuffer {
     fn process_bytes(&mut self, bytes: &[u8]) {
-        for b in bytes {
-            self.parser.advance(&mut self.inner, *b);
+        use termwiz::escape::{Action, ControlCode};
+        let actions = self.parser.parse_as_vec(bytes);
+        for action in actions {
+            match action {
+                Action::Print(c) => self.push_character(c),
+                Action::PrintString(s) => {
+                    self.push_str(&s);
+                }
+                Action::Control(ControlCode::LineFeed) => {
+                    self.push_character('\n')
+                }
+                _ => {}
+            }
```

Comment:
nit, TIOLI: per my reading of the `termwiz` docs, we can save the `Vec` allocation if we wrotee this, instead:
```suggestion
        self.parser.parse(bytes, |action| {
            match action {
                Action::Print(c) => self.push_character(c),
                Action::PrintString(s) => {
                    self.push_str(&s);
                }
                Action::Control(ControlCode::LineFeed) => {
                    self.push_character('\n')
                }
                _ => {}
            }
        });
```

Reply:
oh, ugh. i think i know a way to work around that, gimme a sec?

Reply:
@gjcolombo i've opened PR #631 with the way I would refactor this code to allow using `Parser::parse` rather than calling `parse_as_vec`...but then, when I came back to post this, I re-read your comment and saw you mention that you thought the current approach was nicer than having to have a separate `Inner` struct, which I hadn't remembered you saying earlier. So, we can just close that PR --- my bad.